### PR TITLE
Deprecating HIDDEN repo and replacing with 'origin' repos

### PR DESCRIPTION
**Pull request by breillyr**
Replaced all instances of HIDDEN repo dependence with origin repos based on each existing repo.
If the origin repo does not exist it will be automatically created during those operations.

### CODE REVIEW

File: dockpulp/__init__.py [hunk 1]
```diff
@@ -1961,16 +1981,18 @@ def syncRepo(self, env=None, repo=None, config_file=DEFAULT_CONFIG_FILE,
         manifests = list(set(newmanifests) - set(oldmanifests))
         manifests.sort()
 
-        # Need to maintain HIDDEN
-        if repo != HIDDEN and (imgs or manifests):
+        # Need to maintain origin repo
+        self.createOriginRepo(origin_repo)
+
+        if imgs or manifests:
```
Comment (midnightercz) [inline dockpulp/__init__.py]: this is not related to this PR. But for schema2 you will also have to check for manifest lists and tags. Then it makes sense to copy everything everytime and leave filtering of already copied units to pulp.

File: dockpulp/__init__.py [hunk 2]
```diff
@@ -67,6 +67,7 @@
 DEFAULT_DISTRIBUTORS_FILE = '/etc/dockpulpdistributors.json'
 DEFAULT_DISTRIBUTIONS_FILE = '/etc/dockpulpdistributions.json'
 PREFIX = 'redhat-'
+ORIGIN = 'origin-'
```
[the remark this answers is not in the document]
Comment (midnightercz) [reply dockpulp/__init__.py]: also, we internally use temp- instead of origin-

File: dockpulp/__init__.py [hunk 3]
```diff
@@ -1153,10 +1154,17 @@ def crane(self, repos=[], wait=True, skip=False, force_refresh=False):
         else:
             return tasks
 
+    def createOriginRepo(self, repo_id):
+        try:
+            self.listRepos(repos=repo_id)
+        except errors.DockPulpError:
+            # only create origin repo if it does not exist
+            self.createRepo(repo_id, None, distributors=False, prefix_with=ORIGIN, origin=True)
```
[the remark this answers is not in the document]
Comment (midnightercz) [reply dockpulp/__init__.py]: If we consider purpose of origin repos e.i. avoid have orphans in pulp db, old content is available in HIDDEN. Only new content will be in origin- repos. So it's probably not necessary to copy everything over.

[the remark this answers is not in the document]
Comment (lcarva) [reply dockpulp/__init__.py]: Good point. Getting rid of the existing HIDDEN repo can be done at a later time if that's desired.

File: dockpulp/__init__.py [hunk 4]
```diff
@@ -1040,10 +1041,28 @@ def cleanUploadRequests(self):
         for upload in uploads:
             self._deleteUploadRequest(upload)
 
-    def copy(self, drepo, img, source=HIDDEN):
+    def copy(self, drepo, img, source=None):
         """Copy an image from one repo to another."""
         if img.startswith("sha256:"):
 
+            if source is None:
+                search_data = json.dumps({
```
Comment (lcarva) [inline dockpulp/__init__.py]: It would be really neat if this was a helper method:
```python
def _find_content_source(content_id, content_filter, content_types):
  # content_filter: 'digest' or 'image_id'
  # content_types: (V2_C_TYPE, V2_LIST, V2_BLOB, V2_TAG) or (V1_C_TYPE,)
``` 

It would make it much easier to read this method.

File: dockpulp/__init__.py [hunk 5]
```diff
@@ -1153,10 +1188,18 @@ def crane(self, repos=[], wait=True, skip=False, force_refresh=False):
         else:
             return tasks
 
+    def createOriginRepo(self, repo_id):
+        try:
+            self.listRepos(repos=repo_id)
```
Comment (lcarva) [inline dockpulp/__init__.py]: Should prefix be added here before querying it? `self.listRepos(repos=ORIGIN_PREFIX + repo_id)`
Or better, do it conditionally if it's not already done. I say this because this method is public and having the caller add prefix seems to be an odd requirement.

File: dockpulp/__init__.py [hunk 6]
```diff
@@ -1354,31 +1403,30 @@ def getAllRepoIDs(self):
...
+            img = self._post('/pulp/api/v2/content/units/%s/search/' % V1_C_TYPE,
                              data=json.dumps(data))[0]
         except IndexError:
             log.info('missing parent layer %s', iid)
             log.info('skipping layer')
             return parents
-        img['metadata'].setdefault('parent_id', None)
-        par = img['metadata']['parent_id']
+        img.setdefault('parent_id', None)
+        par = img['parent_id']
```
Comment (lcarva) [inline dockpulp/__init__.py]: `setdefault` actually returns the value, this could be `par = img.setdefault('parent_id', None)`

File: dockpulp/__init__.py [hunk 7]
```diff
@@ -1953,6 +2004,11 @@ def syncRepo(self, env=None, repo=None, config_file=DEFAULT_CONFIG_FILE,
         else:
             oldmanifests = repoinfo['manifests'].keys()
 
+        if len(repoinfo['manifest_lists'].keys()) == 0:
```
Comment (lcarva) [inline dockpulp/__init__.py]: This is not needed. `dict.keys()` will return an empty list. `oldmanifestlists = repoinfo['manifests'].keys()` should handle both cases.

File: dockpulp/__init__.py [hunk 8]
```diff
@@ -2127,15 +2184,24 @@ def upload(self, image):
                 'filename': os.path.basename(image)
             }
         }
-        log.info('adding %s to %s' % (iid, HIDDEN))
+        if drepo != HIDDEN and not drepo.startswith(ORIGIN_PREFIX):
+            origin_drepo = self.getOriginPrefix() + drepo
+        else:
+            origin_drepo = drepo
+        self.createOriginRepo(drepo)
```
Comment (lcarva) [inline dockpulp/__init__.py]: Wouldn't this create an origin repo for hidden repo in some cases?

File: dockpulp/__init__.py [hunk 9]
```diff
@@ -2127,15 +2184,24 @@ def upload(self, image):
...
-            '/pulp/api/v2/repositories/%s/actions/import_upload/' % HIDDEN,
+            '/pulp/api/v2/repositories/%s/actions/import_upload/' % origin_drepo,
             data=json.dumps(data))
         timer = max(self.timeout, (size / mb) * 2)  # wait 2 seconds per megabyte, or timeout,
         self.watch(tid, timeout=timer)  # whichever is greater
         self._deleteUploadRequest(rid)
+        # use filter to copy all new images
+        pulp_filter = {'unit': {
+            '$or': [{'image_id': img} for img in newimgs]}}
+        self.copy(drepo, source=origin_drepo, filters=pulp_filter, v1=True, v2=False)
```
Comment (lcarva) [inline dockpulp/__init__.py]: `copy_filters` ?

File: dockpulp/__init__.py [hunk 10]
```diff
@@ -2085,13 +2140,15 @@ def updateAutoPublish(self, rid, dist_id, auto_publish, watch=True):
             self.watch(tid)
         return tid
 
-    def upload(self, image):
+    def upload(self, image, drepo=HIDDEN):
```
Comment (lcarva) [inline dockpulp/__init__.py]: I think this is the only method that we can't automatically transition to origin repo... 
I'm wondering if this should be a breaking change, i.e. always require `drepo` parameter.

Although, I guess it's not too bad that it's still using `HIDDEN`.

File: dockpulp/__init__.py (hunk 5 above)
Comment (rcerven) [reply dockpulp/__init__.py]: it seems to be query for any repo, even without origin, and crates origin only if repo doesn't exists

File: dockpulp/__init__.py [hunk 11]
```diff
@@ -1294,6 +1337,12 @@ def createRepo(self, repo_id, url, registry_id=None, desc=None, title=None, prot
                     continue
         else:
             stuff['distributors'] = []
+
+        if not is_origin and repo_id != HIDDEN and repo_id != SIGSTORE:
+            # want to create origin- repo for every new repo
+            # do this at the end in case of errors
+            self.createOriginRepo(ORIGIN_PREFIX + repo_id)
```
Comment (rcerven) [inline dockpulp/__init__.py]: can't we just raise error in createOriginRepo, and not call createRepo again,
so it if doesn't raise it can just continue with creation?

File: dockpulp/cli.py [hunk 12]
```diff
@@ -874,6 +873,15 @@ def do_sync(bopts, bargs, parser):
             log.info(manifest)
     log.info('')
 
+    log.info('synced manifest lists:')
+
+    if len(manifest_lists) == 0:
```
Comment (rcerven) [inline dockpulp/cli.py]: can be just "if manifest_lists" since it will be empty list if it is empty

File: dockpulp/__init__.py (hunk 8 above)
Comment (breillyr) [reply dockpulp/__init__.py]: It wouldn't have if createOriginRepo did not automatically prepend w/ the origin prefix (as it would have checked if 'redhat-everything' exists and then stopped there), but now that it does this must be changed, yes.

File: dockpulp/__init__.py (hunk 10 above)
Comment (breillyr) [reply dockpulp/__init__.py]: I'd rather not make it a breaking change as I don't think there is much of a need. Other tools will have to be updated to match this and I don't want the functionality to break in the meantime. Once they're updated I can remove it at that point and require repo input.

File: dockpulp/__init__.py (hunk 11 above)
Comment (breillyr) [reply dockpulp/__init__.py]: What we want to avoid here is the origin repo being successfully created while the actual repo fails - the error the comment is talking about there is a parsing error for the original repo. Essentially it's the sanity check before we commit to creating the origin repo - if we fail before we even make the call to Pulp, the origin repo should not be created with that faulty information.

File: dockpulp/__init__.py [hunk 13]
```diff
@@ -1058,11 +1083,14 @@ def copy(self, drepo, img, source=HIDDEN):
             }
 
         else:
+            content_types = [V1_C_TYPE]
```
[the remark this answers is not in the document]
Comment (lcarva) [reply dockpulp/__init__.py]: Isn't `docker_tag` v2 content that is automatically generated for manifests?

[the remark this answers is not in the document]
Comment (twaugh) [reply dockpulp/__init__.py]: Well sure, but you don't reference a docker_tag unit by the digest it refers to.
